### ISSUE TITLE
readdir performance regression since 1.84

### ISSUE THREAD

**Issue by gaul** [s3fs-fuse/s3fs-fuse#935]
Using the goofys benchmark script, I observe much slower `ls_files` performance in master than 1.84 (results are time, lower is better):

1.84:

```
$ ~/work/go-workspace/src/github.com/kahing/goofys/bench/bench.sh '/home/gaul/work/s3fs-fuse/src/s3fs gaultest-us-west-1 /home/gaul/mnt /home/gaul/mnt ls
ls_files 2.913
ls_files 0.985
ls_files 0.829
ls_files 0.852
ls_files 0.808
ls_files 0.811
ls_files 1.093
ls_files 0.959
ls_files 0.933
ls_files 0.840
```

master:

```
$ ~/work/go-workspace/src/github.com/kahing/goofys/bench/bench.sh '/home/gaul/work/s3fs-fuse/src/s3fs gaultest-us-west-1 /home/gaul/mnt' /home/gaul/mnt ls
ls_files 16.186
ls_files 9.720
ls_files 11.963
ls_files 12.418
ls_files 10.156
ls_files 10.072
ls_files 9.872
ls_files 9.751
ls_files 9.615
ls_files 10.418
```

We should analyze this before releasing 1.85.  The script reveals other interesting results including improvements to bulk read and write performance.

**Comment by ggtakec** [comment]
@gaul Thanks for the report.
I will concentrate and check it this week.

**Comment by gaul** [comment]
The benchmark script may have an issue.  I observe a 25% improvement when running the same test manually:

without 46d79c5bc275c1df8cdd4e2ca0884efc68247e77:
```
$ time ls mnt/dir/ | wc -l
1000

real    0m8.109s
user    0m0.001s
sys     0m0.003s
```

with 46d79c5bc275c1df8cdd4e2ca0884efc68247e77:
```
$ time ls mnt/dir/ | wc -l
1000

real    0m5.715s
user    0m0.007s
sys     0m0.005s
```

**Comment by gaul** [comment]
It appears 46d79c5bc275c1df8cdd4e2ca0884efc68247e77 increases CPU usage partially from `malloc_trim`.  Still investigating...

**Comment by gaul** [comment]
I suspect that s3fs is not reusing HTTPS connections and the CPU overhead comes from the initial handshake.

**Comment by ggtakec** [comment]
@gaul I confirmed that the session was not reused and HTTPS(TLS) handshake was running.
I will immediately check the contents of the code.

**Comment by ggtakec** [comment]
The session with S3 has been disconnected from the server in around 5 seconds and it is CLOSE_WAIT.
Now I can not control this with the CURL option.
(I tried the CURLOPT_TCP_KEEPALIVE option etc.)
I will examine it in succession.

**Comment by ggtakec** [comment]
One reason why the socket is closed was that we received 400 errors in the HEAD request(ex. ls <dir>).
In case of 400 error, "Connection: close" is received and closed with libcurl.
By specifying CURLOPT_KEEP_SENDING_ON_ERROR by curl_easy_setopt, closing at HEAD request seems to be avoidable.
I'll continue to check about multipart upload.

**Comment by ggtakec** [comment]
I merged #965 .

It does not set the CURLOPT_KEEP_SENDING_ON_ERROR option.
So, renegotiation of SSL may occur at HEAD request.

However, I reviewed the pool of curl handles with that PR and reviewed the processing of multipart request.
As a result, renegotiation of SSL does not occur in multipart POST/GET.

Please see #965 comments for detail.

**Comment by gaul** [comment]
Fixed by #965.
